src/zenml/deployers/serving/runtime.py

```diff
@@ -0,0 +1,374 @@
+    enforce_size_limits: bool = True,
+    max_output_size_mb: int = 1,
+) -> Dict[str, Any]:
+    """Process outputs using fast path when available, slow path as fallback.
+
+    Args:
+        runtime_outputs: In-memory outputs from runtime context (fast path)
+        run: Pipeline run response for artifact loading (slow path)
+        enforce_size_limits: Whether to enforce size limits (disable for in-memory mode)
+        max_output_size_mb: Maximum output size in MB
+
+    Returns:
+        Processed outputs ready for JSON response
+    """
+    if runtime_outputs:
+        return _process_runtime_outputs(
+            runtime_outputs, enforce_size_limits, max_output_size_mb
+        )
+
+    logger.debug("Using slow artifact loading fallback")
+
+    return _process_artifact_outputs(run)
+
+
+def _process_runtime_outputs(
+    runtime_outputs: Dict[str, Dict[str, Any]],
+    enforce_size_limits: bool,
+    max_output_size_mb: int,
+) -> Dict[str, Any]:
+    """Process in-memory outputs with optional size limits."""
+    return {
+        f"{step_name}.{output_name}": _serialize_output(
+            value, enforce_size_limits, max_output_size_mb
+        )
+        for step_name, step_outputs in runtime_outputs.items()
+        for output_name, value in step_outputs.items()
+    }
+
+
+def _serialize_output(
+    value: Any, enforce_size_limits: bool, max_output_size_mb: int
+) -> Any:
+    """Serialize a single output value with error handling."""
+    try:
+        serialized = _make_json_safe(value)
+
+        if not enforce_size_limits:
+            return serialized
+
+        # Check size limits only if enforced
+        max_size_bytes = max(1, min(max_output_size_mb, 100)) * 1024 * 1024
+        if isinstance(serialized, str) and len(serialized) > max_size_bytes:
+            return {
+                "data_too_large": True,
+                "size_estimate": f"{len(serialized) // 1024}KB",
+                "max_size_mb": max_size_bytes // (1024 * 1024),
+                "type": type(value).__name__,
+                "note": "Use artifact loading endpoint for large outputs",
+            }
+
+        return serialized
+
+    except Exception:
+        return {
+            "serialization_failed": True,
+            "type": type(value).__name__,
+            "note": "Use artifact loading endpoint for this output",
+        }
+
+
+def _process_artifact_outputs(run: PipelineRunResponse) -> Dict[str, Any]:
+    """Load outputs from artifacts and serialize them safely.
+
+    Args:
+        run: Pipeline run response to iterate step outputs.
+
+    Returns:
+        Mapping from "step.output" to serialized values.
+    """
+    from zenml.artifacts.utils import load_artifact_from_response
+
+    outputs: Dict[str, Any] = {}
+    for step_name, step_run in (run.steps or {}).items():
+        if not step_run or not step_run.outputs:
+            continue
+
+        for output_name, artifacts in step_run.outputs.items():
```

Comment:
You probably want `step_run.regular_outputs` here.

Reply:
not used anymore